### PR TITLE
s/a/notify: implement protocol v5 with metadata tagging parser

### PR DESCRIPTION
**Pull request by olivercalder**
This PR is based on #15130 and #15131.

Implement version 5 of the apparmor notification protocol, as defined here: https://docs.google.com/document/d/1_WvEM9Qi2Je2Vwulzv5TzHwLJ8ld0W8gTNESZZd9VsY/edit?tab=t.0#heading=h.q2d8zdyargyy

This work is tracked internally by https://warthogs.atlassian.net/browse/SNAPDENG-32517

### CODE REVIEW

File: sandbox/apparmor/notify/message.go [hunk 1]
```diff
@@ -488,12 +524,62 @@ func (msg *MsgNotificationFile) UnmarshalBinary(data []byte) error {
...
+	TagCount       uint32
+	TagOffset      uint32
+}
+
+func (msg *MsgNotificationFile) unmarshalTags(data []byte) error {
+	// Unpack fixed-size elements to get tag metadata.
+	buf := bytes.NewBuffer(data)
+	var raw msgNotificationFileKernelWithTags
+	order := arch.Endian() // ioctl messages are native byte order, verify endianness if using for other messages
+	if err := binary.Read(buf, order, &raw); err != nil {
```
Comment (zyga) [inline sandbox/apparmor/notify/message.go:538]: Isn't this just a slice? No need for binary.Read for bytes.

File: sandbox/apparmor/notify/message.go [hunk 2]
```diff
@@ -488,12 +524,62 @@ func (msg *MsgNotificationFile) UnmarshalBinary(data []byte) error {
...
 
+type tagsetHeader struct {
+	PermissionMask uint32
+	TagCount       uint32
+	TagOffset      uint32
+}
+
+func (msg *MsgNotificationFile) unmarshalTags(data []byte) error {
+	// Unpack fixed-size elements to get tag metadata.
+	buf := bytes.NewBuffer(data)
```
Comment (zyga) [inline sandbox/apparmor/notify/message.go:535]: bytes.NewReader is cheaper

File: sandbox/apparmor/notify/message.go [hunk 3]
```diff
@@ -488,12 +524,62 @@ func (msg *MsgNotificationFile) UnmarshalBinary(data []byte) error {
...
+		return fmt.Errorf("cannot unpack tagset metadata: %v", err)
+	}
+
+	if raw.Tags == 0 || raw.TagsetsCount == 0 {
+		return nil
+	}
+
+	// Unpack the tagset headers.
+	tagsets := make(map[uint32][]string, raw.TagsetsCount)
+	hdrBuf := bytes.NewBuffer(data[raw.Tags:])
```
Comment (zyga) [inline sandbox/apparmor/notify/message.go:548]: Why do we need a new buffer for this? Just continue to read from the previous reader?

File: sandbox/apparmor/notify/message_test.go [hunk 4]
```diff
@@ -481,11 +481,110 @@ func (s *messageSuite) TestMsgNotificationFileMarshalUnmarshalBinary(c *C) {
...
+	c.Assert(err, IsNil)
+	c.Assert(roundTripMsg, DeepEquals, expected)
+}
+
+func (s *messageSuite) TestMsgNotificationFileUnmarshalBinaryV5(c *C) {
+	if arch.Endian() == binary.BigEndian {
+		c.Skip("test only written for little-endian architectures")
+	}
+	// Notification for accessing a the /root/.ssh/ file.
+	bytes := []byte{
```
Comment (zyga) [inline sandbox/apparmor/notify/message_test.go:502]: Could you please add this as a binary file to `testdata/`. I will add a imhex profile for reading it. I think it's more useful than an annotated byte slice in the code.

File: sandbox/apparmor/notify/strings.go [hunk 5]
```diff
@@ -39,6 +41,57 @@ func (sp *stringPacker) PackString(s string) uint32 {
...
+// By convention, there is an additional \0 byte after the end of each tagset,
+// but in the future, tagsets may overlap to save space, so this should not be
+// relied upon, so we do not include it here.
+func (sp *stringPacker) PackTagsets(ts map[uint32][]string) uint32 {
+	if len(ts) == 0 {
+		return 0
+	}
+	headers := make([]tagsetHeader, len(ts))
+	i := 0
+	for perm, tags := range ts {
```
Comment (zyga) [inline sandbox/apparmor/notify/strings.go:67]: This is not deterministic. Can we make it so?

File: sandbox/apparmor/notify/strings.go [hunk 6]
```diff
@@ -78,3 +131,28 @@ func (su *stringUnpacker) UnpackString(offset uint32) (string, error) {
...
+	}
+	strs := make([]string, n)
+	for i := uint32(0); i < n; i++ {
+		tmp := su.Bytes[offset:]
+		idx := bytes.IndexByte(tmp, 0)
+		if idx < 0 {
+			return nil, fmt.Errorf("unterminated string at address %d", offset)
+		}
+		strs[i] = string(tmp[:idx])
+		offset += uint32(idx + 1) // advance offset to start of next string
```
Comment (zyga) [inline sandbox/apparmor/notify/strings.go:155]: I think this is more precise, no?
```suggestion
		offset += uint32(idx) + 1 // advance offset to start of next string
```

File: sandbox/apparmor/notify/version.go [hunk 7]
```diff
@@ -41,7 +55,34 @@ var (
...
+			if osutil.IsDirectory(protocolVersionsDir) && !notifyVersionFileExists(3) {
+				return false
+			}
+			return true
+		},
+		5: func() bool {
+			if !SupportAvailable() {
+				return false
+			}
+			if !notifyVersionFileExists(5) {
```
Comment (zyga) [inline sandbox/apparmor/notify/version.go:71]: Do we need to check if `protocolVersionsDir` exits like we do for 3 above?

File: sandbox/apparmor/notify/version.go [hunk 8]
```diff
@@ -41,7 +55,34 @@ var (
...
+			// XXX: apparmor.KernelFeatures() already has this information, but
+			// we can't import apparmor here since that would be circular.
+			data, err := os.ReadFile(protocolFeaturesPath)
+			if err != nil {
+				return false
+			}
+			features := strings.Fields(string(data))
+			if !strutil.ListContains(features, "tags") {
+				return false
+			}
```
[the remark this answers is not in the document]
Comment (zyga) [reply sandbox/apparmor/notify/version.go:83]: I agree, should we discuss restructuring so that it can happen?

File: sandbox/apparmor/notify/version.go (hunk 7 above)
Comment (olivercalder) [reply sandbox/apparmor/notify/version.go:71]: If the directory doesn't exist, version 5 is not supported. Version 3 is the special case, if there's no directory at all we assume support for version 3.

File: sandbox/apparmor/notify/message_test.go (hunk 4 above)
Comment (olivercalder) [reply sandbox/apparmor/notify/message_test.go:502]: I can do this, but I think it may make it harder to read/modify the unit tests in the future, as versions are added. We'll need to know where to look, and load the binary files in the right place. I suppose these are totally reasonable things to do though :)

File: sandbox/apparmor/notify/version.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (olivercalder) [reply sandbox/apparmor/notify/version.go:83]: `sandbox/apparmor` only imports this package so it can call `notify.SupportAvailable()` (which checks the existence of the notify socket). So a much simpler solution could be to just move the `SupportAvailable` check to the `apparmor` package, and define the `NotifySocketPath` in `apparmor` as well, and import those, plus the version support checks, into `sandbox/apparmor/notify`.

File: sandbox/apparmor/notify/message.go (hunk 3 above)
Comment (olivercalder) [reply sandbox/apparmor/notify/message.go:548]: Tagset headers might not start immediately after the base structure, they could be anywhere in the following data section, so this builds a new buffer starting at where the tagset headers begin.

File: sandbox/apparmor/notify/message_test.go (hunk 4 above)
Comment (zyga) [reply sandbox/apparmor/notify/message_test.go:502]: I can try to do a quick pass on that tomorrow. I think we will realistically start parsing regression test data or other real examples and just having all of that as annotated text is not scalable IMO.

File: sandbox/apparmor/notify/version.go (hunk 7 above)
Comment (zyga) [reply sandbox/apparmor/notify/version.go]: Can you leave a comment in the code please.

**Comment by zyga** [comment]
@olivercalder can you rebase it once the prerequisite lands please.

**Comment by zyga** [comment]
@olivercalder shall I try to review without rebasing?

File: sandbox/apparmor/notify/strings.go [hunk 9]
```diff
@@ -4,6 +4,9 @@ import (
 	"bytes"
 	"encoding/binary"
 	"fmt"
+	"slices"
```
Comment (olivercalder) [inline sandbox/apparmor/notify/strings.go:7]: I don't think this is available in the go version available in focal :(

Comment (bboozzoo) [reply sandbox/apparmor/notify/strings.go:7]: slices is go 1.21+

File: sandbox/apparmor/notify/version.go [hunk 10]
```diff
@@ -30,7 +33,7 @@ var (
 	// versions holds the notification protocols snapd supports, in order of
 	// preference. If the first version is supported, try to use it, else try
 	// the next version, etc.
-	versions = []ProtocolVersion{3}
+	versions = []ProtocolVersion{5, 3}
```
Comment (bboozzoo) [inline sandbox/apparmor/notify/version.go:36]: out of curiosity, what happened to 4?

File: sandbox/apparmor/notify/message.go [hunk 11]
```diff
@@ -488,12 +524,62 @@ func (msg *MsgNotificationFile) UnmarshalBinary(data []byte) error {
...
-	var raw msgNotificationFileKernel
-	packer := newStringPacker(raw)
+	var raw msgNotificationFileKernelWithTags // prepare the superset of the msgNotificationFileKernel
+	var ptr any
+	if msg.Version < 5 {
+		ptr = &raw.msgNotificationFileKernelBase
+	} else {
+		ptr = &raw
+	}
+	packer := newStringPacker(ptr)
```
Comment (bboozzoo) [inline sandbox/apparmor/notify/message.go:582]: seems like newStringPacker only takes the struct in order to compute its binary size, if you change the API to `func newStringPacker(size int) ..`, you'll no longer need to pass `any`.

File: sandbox/apparmor/notify/message.go [hunk 12]
```diff
@@ -488,12 +524,62 @@ func (msg *MsgNotificationFile) UnmarshalBinary(data []byte) error {
...
+		tagsets[header.PermissionMask] = tags
+	}
+
+	msg.Tagsets = tagsets
+
+	return nil
+}
+
+// It should not be necessary to marshal MsgNotificationFile structs outside of
+// test code.
```
[the remark this answers is not in the document]
Comment (olivercalder) [reply sandbox/apparmor/notify/message.go:570]: The problem is `sandbox/apparmor/notify/listener/listener_test.go` needs to marshal `MsgNotificationFile`s as well, at least as the tests are written now, otherwise yes definitely

File: sandbox/apparmor/notify/strings.go [hunk 13]
```diff
@@ -39,6 +42,69 @@ func (sp *stringPacker) PackString(s string) uint32 {
...
+//
+// The tagsets themselves may occur before or after the tagset headers, and
+// need not be contiguous, either between tagsets or with the tagset headers.
+// All the tags in any given tagset must be contiguous, however.
+//
+// For code simplicity, we encode all the tagsets first, then the headers.
+// By convention, there is an additional \0 byte after the end of each tagset,
+// but in the future, tagsets may overlap to save space, so this should not be
+// relied upon, so we do not include it here.
+func (sp *stringPacker) PackTagsets(ts map[uint32][]string) uint32 {
```
[the remark this answers is not in the document]
Comment (olivercalder) [reply sandbox/apparmor/notify/strings.go:62]: It's not used outside of `MsgNotificationFile.MarshalBinary()`, so I believe it could be at least private. But since `MsgNotificationFile.MarshalBinary()` is called from `s/a/n/listener/listener_test.go`, I don't think it can be defined in test code.

File: sandbox/apparmor/notify/version.go (hunk 10 above)
Comment (olivercalder) [reply sandbox/apparmor/notify/version.go:36]: Great question, I'll ask JJ :)

File: sandbox/apparmor/notify/message.go (hunk 11 above)
Comment (olivercalder) [reply sandbox/apparmor/notify/message.go:582]: True, but then the caller will need to use `binary.Size(theStruct)` or otherwise compute the size of the struct correctly. OTOH this pointer gymnastics is a bit much :) so I'm open to it

Comment (olivercalder) [reply sandbox/apparmor/notify/message.go:582]: `binary.Size` takes `any` too, so I don't really think it would be an improvement, but I'm happy to be overruled here

File: sandbox/apparmor/notify/version.go [hunk 14]
```diff
@@ -42,8 +45,34 @@ var (
 	// the next version in the list.
 	versionLikelySupportedChecks = map[ProtocolVersion]func() bool{
 		3: func() bool {
```
Comment (pedronis) [inline sandbox/apparmor/notify/version.go:47]: a bit surprised but coverage seems to indicated that the functions in this map are not tested

File: sandbox/apparmor/notify/strings.go [hunk 15]
```diff
@@ -39,6 +42,69 @@ func (sp *stringPacker) PackString(s string) uint32 {
 	return offset + uint32(sp.baseOffset)
 }
 
+// packTagsets computes the layout of the tagsets encoded in an apparmor message.
```
Comment (pedronis) [inline sandbox/apparmor/notify/strings.go:45]: this doc is a bit confusing as the function actually also packs the tagsets, also it doesn't clearly describe how ts is organized in the input, are the keys permissions?

File: sandbox/apparmor/notify/strings.go [hunk 16]
```diff
@@ -39,6 +42,69 @@ func (sp *stringPacker) PackString(s string) uint32 {
...
+// need not be contiguous, either between tagsets or with the tagset headers.
+// All the tags in any given tagset must be contiguous, however.
+//
+// For code simplicity, we encode all the tagsets first, then the headers.
+// By convention, there is an additional \0 byte after the end of each tagset,
+// but in the future, tagsets may overlap to save space, so this should not be
+// relied upon, so we do not include it here.
+func (sp *stringPacker) packTagsets(ts map[uint32][]string) uint32 {
+	if len(ts) == 0 {
+		return 0
```
Comment (pedronis) [inline sandbox/apparmor/notify/strings.go:64]: this is not tested

File: sandbox/apparmor/notify/strings.go [hunk 17]
```diff
@@ -39,6 +42,69 @@ func (sp *stringPacker) PackString(s string) uint32 {
...
+	}
+	// TODO: use slices.Sort() once we're on go 1.21+
+	sort.Slice(perms, func(i, j int) bool {
+		return perms[i] < perms[j]
+	})
+
+	for _, perm := range perms {
+		tags := ts[perm]
+		if len(tags) == 0 {
+			continue
```
Comment (pedronis) [inline sandbox/apparmor/notify/strings.go:83]: same, not tested

File: sandbox/apparmor/notify/message.go [hunk 18]
```diff
@@ -488,12 +524,62 @@ func (msg *MsgNotificationFile) UnmarshalBinary(data []byte) error {
...
+	}
+
+	if raw.Tags == 0 || raw.TagsetsCount == 0 {
+		return nil
+	}
+
+	// Unpack the tagset headers.
+	tagsets := make(map[uint32][]string, raw.TagsetsCount)
+	hdrBuf := bytes.NewReader(data[raw.Tags:])
+	headers := make([]tagsetHeader, raw.TagsetsCount)
```
Comment (pedronis) [inline sandbox/apparmor/notify/message.go:549]: this seems slightly wasteful, you read one header at a time in the main loop below, without needing to copy out all the headers first, unless I'm missing something

**Review by pedronis — COMMENTED** [review]
did a first pass

File: sandbox/apparmor/notify/message.go (hunk 18 above)
Comment (olivercalder) [reply sandbox/apparmor/notify/message.go:549]: I copied them into a `[]tagsetHeader` so that there's a safe way to iterate through them. The tagsets themselves could be at arbitrary places in the data, but the headers are all contiguous and each contains the offset of the corresponding tagset. Is there a way to read the headers from the existing buffer without unsafe casting the `[]byte` to `[]tagsetHeader`?

Comment (pedronis) [reply sandbox/apparmor/notify/message.go:549]: can't you do this?

```
	tagsets := make(map[uint32][]string, raw.TagsetsCount)
	hdrBuf := bytes.NewReader(data[raw.Tags:])
	// Unpack the tags associated with each tagset header.
	unpacker := newStringUnpacker(data)
	for i := 0; i < int(raw.TagsetsCount); i++ {
		var header tagsetHeader
		if err := binary.Read(hdrBuf, order, &header); err != nil {
			return fmt.Errorf("cannot unpack tagset header: %v", err)
		}
		tags, err := unpacker.UnpackStrings(header.TagOffset, header.TagCount)
       ...
 ```

Comment (olivercalder) [reply sandbox/apparmor/notify/message.go:549]: Ah yes, that works, and avoids a heap allocation. Nice!

**Comment by olivercalder** [comment]
Should methods on the `stringPacker` be public or private? The type is private and only used in the `s/a/notify` package, but the existing methods are all public, seems weird to have just `packTagsets` be private.

File: sandbox/apparmor/notify/message.go [hunk 19]
```diff
@@ -429,21 +429,38 @@ func (msg *MsgNotificationOp) UnmarshalBinary(data []byte) error {
...
+type msgNotificationFileKernelBase struct {
 	msgNotificationOpKernel
 	SUID uint32
 	OUID uint32
 	Name uint32
 }
 
+// msgNotificationFileKernelWithTags (protocol version >=5)
+//
+//	struct apparmor_notif_file {
```
[the remark this answers is not in the document]
Comment (olivercalder) [reply sandbox/apparmor/notify/message.go:449]: I asked JJ for a link and he sent a pastebin of the difference between v3 and v5 (https://pastebin.canonical.com/p/6dYf2jTnqG/) and a complete copy of the v5 definitions (https://pastebin.canonical.com/p/HpVFVsdcwH/)

I'm not certain that the v5 protocol is landed in the final place where we can put a permalink in the source code yet.

File: sandbox/apparmor/notify/strings.go [hunk 20]
```diff
@@ -78,3 +155,25 @@ func (su *stringUnpacker) UnpackString(offset uint32) (string, error) {
...
+		return nil, fmt.Errorf("address %d points outside of message body", offset)
+	}
+	strs := make([]string, n)
+	for i := uint32(0); i < n; i++ {
+		tmp := su.Bytes[offset:]
+		idx := bytes.IndexByte(tmp, 0)
+		if idx < 0 {
+			return nil, fmt.Errorf("unterminated string at address %d", offset)
+		}
+		strs[i] = string(tmp[:idx])
```
Comment (bboozzoo) [inline sandbox/apparmor/notify/strings.go:175]: the spec mentions that tags are valid utf8, we could try `utf8.ValidString()` on what we get here if it isn't too wasteful, though the strings should be exactly what we put into the profiles, so maybe it's not really worth it?

a side question, how do we handle file paths passed up from AppArmor wrt encoding? looking at unpackString() we simply assume they are utf8 strings, but in fact a filesystem may use different arbitrary encoding

File: sandbox/apparmor/notify/message_test.go [hunk 21]
```diff
@@ -481,11 +481,336 @@ func (s *messageSuite) TestMsgNotificationFileMarshalUnmarshalBinary(c *C) {
...
+		0x03, 0x00, 0x00, 0x00, // tagset 2 tag count
+		0x84, 0x00, 0x00, 0x00, // tagset 2 starts at +132 into the buffer
+		0x04, 0x00, 0x00, 0x00, // tagset 3 permission mask
+		0x02, 0x00, 0x00, 0x00, // tagset 3 tag count
+		0x80, 0x00, 0x00, 0x00, // tagset 3 starts at +128 into the buffer
+		0x08, 0x00, 0x00, 0x00, // tagset 4 permission mask
+		0x02, 0x00, 0x00, 0x00, // tagset 4 tag count
+		0x84, 0x00, 0x00, 0x00, // tagset 4 starts at +132 into the buffer
+		0x6f, 0x6e, 0x65, 0x00, // "one\0"
+		0x74, 0x77, 0x6f, 0x00, // "two\0"
```
Comment (bboozzoo) [inline sandbox/apparmor/notify/message_test.go:696]: if I'm reading the AA008 spec correctly, last tag in a tagset is followed by additional `\0`, is this still the case? (fwiw, it's transparent for snapd since the parsing part looks at offsets)

Comment (olivercalder) [reply sandbox/apparmor/notify/message_test.go:696]: It is the case by convention, so one could parse tags in a tagset without knowing how many there are. But JJ discussed the possibility of overlapping tagsets which have common tags, in which case they would not have a double `'\0'` at the end. This test is for those edge cases: an empty tagset, and some tagsets which overlap, so there's no extra `'\0'` here.

File: sandbox/apparmor/notify/strings.go [hunk 22]
```diff
@@ -41,11 +41,11 @@ func (sp *stringPacker) PackString(s string) uint32 {
 	return offset + uint32(sp.baseOffset)
 }
 
-// packTagsets computes headers for the given tagsets writes the tagsets and
-// headers into the packer, and returns the offset of the beginning of the
-// first header relative to the start of the structure, captured by baseOffset.
+// packTagsets computes headers for the given tagsets packs the tagsets and
```
Comment (pedronis) [inline sandbox/apparmor/notify/strings.go:44]: is there a missing comma,  "given tagsets, packs the..." ?

**Review by pedronis — APPROVED** [review]
thanks

File: sandbox/apparmor/notify/strings.go (hunk 20 above)
Comment (olivercalder) [reply sandbox/apparmor/notify/strings.go:175]: Since we hard-code the strings, I don't think it's particularly necessary, but I'll .

Regarding the broader `string` vs `OSString` (or whatever it's called in Go), I think it's something that needs to be addressed more broadly across the prompting components. We use `string` everywhere. But from what I can tell, Go doesn't have a separate concept of `string` and `OSString` (to use the Rust type)... it seems `string` can represent arbitrary bytes. And I don't immediately see what about `unpackString` assumes it's valid UTF-8, unless slice indices work on runes instead of bytes??? but I don't think that's the case?? No, string indexing is against bytes: https://go.dev/blog/strings

```
/ unpackString unpacks NUL-terminated string at a given offset into the buffer.
func (su *stringUnpacker) unpackString(offset uint32) (string, error) {
	if offset == 0 {
		return "", nil
	}
	if offset >= uint32(len(su.Bytes)) {
		return "", fmt.Errorf("address %d points outside of message body", offset)
	}
	tmp := su.Bytes[offset:]
	idx := bytes.IndexByte(tmp, 0)
	if idx < 0 {
		return "", fmt.Errorf("unterminated string at address %d", offset)
	}
	return string(tmp[:idx]), nil
}
```

Since `string` can be arbitrary bytes, shouldn't `string(tmp[:idx])` be fine, even if for some reason the string isn't utf-8? And `len(su.Bytes)` is correct, and offset just points to a start byte. So I don't see what wouldn't work for non-utf8 strings. But I don't know the Go intricacies as well as Rust, am I missing something here?

By the way `unpackString()` is used to unpack strings for arbitrary filepaths which triggered AppArmor requests, so that feels like a bigger problem we need to solve if this can't handle non-utf8 strings.

Comment (olivercalder) [reply sandbox/apparmor/notify/strings.go:175]: Tags are now validated to be of a strict set of characters here: https://github.com/canonical/snapd/pull/15199/files#diff-78692e23a83df81d72626c6a2a4cb7d54281f8fbaa33073bf019555fe760500aR120

**Review by bboozzoo — APPROVED** [review]
LGTM

File: sandbox/apparmor/notify/strings.go (hunk 20 above)
Comment (olivercalder) [reply sandbox/apparmor/notify/strings.go:175]: Yeah that's a good idea. And it's the paths/path patterns in the prompts and rules sent over the API which are the concern I think, FTR notices just have the prompt/rule ID so the client can look them up. If go is doing some coercion, hopefully it's symmetrical (that is, if snapd receives a path with messy codepoints, it can convert back to non-UTF-8 after it's received), but I don't know whether that's actually true. Worst case, replies/rules won't match requests if the former have replacement codepoints, which would be very annoying but at least not weaken security by letting requests by. Definitely worth investigating more and fixing. I'll add a jira card to my backlog. Thanks!

**Comment by olivercalder** [comment]
Since JJ is planning to roll the restarts changes (formerly v7) into v5 as well, I'm going to mark v5 as manually disabled for now, in case we cut snapd 2.69 before the rest of the v7->v5 changes land, so snapd doesn't incorrectly report that it supports the full v5 with restarts. Then I can open new PRs for the restarts work, based on this PR.

**Comment by olivercalder** [comment]
I've manually squashed all the commits except the last one which manually disables protocol v5. Please DO NOT SQUASH, just rebase merge, so that the v5 disable commit can be easily reverted.
